src/main/java/bio/terra/service/filedata/FileMetadataUtils.java

```diff
@@ -196,17 +195,23 @@ public static Md5ValidationResult validateFileMd5ForIngest(
   public static List<FileModel> toFileModel(
       List<FireStoreDirectoryEntry> directoryEntries,
       List<FireStoreFile> files,
-      String collectionId) {
-    if (directoryEntries.size() != files.size()) {
+      String collectionId,
+      boolean enforceMatchingDirectoryEntries) {
+    if (enforceMatchingDirectoryEntries && (directoryEntries.size() != files.size())) {
       throw new FileSystemExecutionException("List sizes should be identical");
     }
 
-    return IntStream.range(0, files.size())
-        .mapToObj(
-            i -> {
-              FireStoreFile file = files.get(i);
-              FireStoreDirectoryEntry entry = directoryEntries.get(i);
-
+    return files.stream()
+        .map(
+            file -> {
+              FireStoreDirectoryEntry entry =
+                  directoryEntries.stream()
+                      .filter(e -> e.getFileId().equals(file.getFileId()))
+                      .findFirst()
+                      .orElseThrow(
+                          () ->
+                              new FileSystemExecutionException(
+                                  "There should be a matching directory entry for every file."));
```

Comment:
why add this exception? and do we only want to throw it if the enforceMatchingDirectoryEntries flag is true?

Reply:
No, the enforceMatchingDirectoryEntries only applies to directory entries that are missing the underlying file metadata entries. 
Why?
- We use the directory entry to look up the file entry. If this file entry exists, then there must be a matching directory entry. 
<img width="1039" alt="image" src="https://github.com/user-attachments/assets/fe5d4303-489d-46d5-8a87-089765b42580">

However, (and this is the case we're accounting for) there could exist a directory entry and then we go to look up the file, no file is found.